cwl/src/main/scala/cwl/WorkflowStepInput.scala

```diff
@@ -86,34 +108,47 @@ case class WorkflowStepInput(
 object WorkflowStepInput {
   type InputSource = String :+: Array[String] :+: CNil
 
-  def determineType(stepInput: WorkflowStepInput,
-                    outputTypeMap: Map[String, WomType],
-                    expectedType: Option[MyriadInputType],
-                    isScattered: Boolean)(implicit parentName: ParentName): Checked[WomType] = {
+  implicit class EnhancedStepInputMap[A](val map: Map[WorkflowStepInput, A]) extends AnyVal {
+    def asIdentifierMap(implicit parentName: ParentName): Map[String, A] = {
+      map.map({ case (stepInput, value) => stepInput.parsedId -> value })
+    }
+  }
 
+  def determineValueFromType(mergedSourcesType: Option[WomType],
+                             expectedType: Option[MyriadInputType],
+                             isScattered: Boolean): Checked[WomType] = {
     val expectedTypeAsWom: Option[WomType] = expectedType.map(_.fold(MyriadInputTypeToWomType))
-    (isScattered, expectedTypeAsWom, stepInput.effectiveLinkMerge) match {
 
-      //If scattering over this variable, we expect an array of the sink type
-      case (true, Some(tpe), _) => WomArrayType(tpe).asRight
+    expectedTypeAsWom.getOrElse(WomStringType).asRight
+  }
 
-      //If sink parameter is an array, we must frame the input as an array
-      case (false, Some(array: WomArrayType), LinkMergeMethod.MergeNested) =>
-        array.asRight
+  def determineMergeType(sources: Map[String, WomType],
+                         linkMerge: Option[LinkMergeMethod],
+                         expectedType: Option[MyriadInputType]): Checked[WomType] = {
 
-      //If sink parameter is an array and merge_flattened is used, must validate input & output types are equivalent before proceeding
-      case (false, Some(array@WomArrayType(tpe)), LinkMergeMethod.MergeFlattened)  =>
-        //Collect the upstream outputs and their types
-        stepInput.validatedSourceTypes(outputTypeMap) flatMap {
-          case map  if typesToItemMatch(map.values, tpe) =>  array.asRight
-          case map => (s"could not verify that types $map and the items type of the run's InputArraySchema $tpe were compatible" |> NonEmptyList.one).asLeft
-        }
+    val expectedTypeAsWom: Option[WomType] = expectedType.map(_.fold(MyriadInputTypeToWomType))
+    
+    (sources.toList, expectedTypeAsWom, linkMerge) match {
```

Comment:
Whew!  I knew we weren't finished w/ this logic.  Seems like it's much improved now.